### PR TITLE
use of checkers on `integration-cli/docker_cli_start_test.go`

### PR DESCRIPTION
**Pull request by WeiZhang555**
Part of #16756

Use c.Assert instead of condition judgement in
integration-cli/docker_cli_start_test.go

Signed-off-by: Zhang Wei zhangwei555@huawei.com

### CODE REVIEW

File: integration-cli/docker_cli_start_test.go [hunk 1]
```diff
@@ -15,9 +16,9 @@ func (s *DockerSuite) TestStartAttachReturnsOnError(c *check.C) {
 	dockerCmd(c, "wait", "test")
 
 	// Expect this to fail because the above container is stopped, this is what we want
-	if _, _, err := dockerCmdWithError("run", "-d", "--name", "test2", "--link", "test:test", "busybox"); err == nil {
-		c.Fatal("Expected error but got none")
-	}
+	_, _, err := dockerCmdWithError("run", "-d", "--name", "test2", "--link", "test:test", "busybox")
+	// Expected error but got none
```
Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: You can even remove this one I guess, it doesn't add much information :stuck_out_tongue_closed_eyes:

File: integration-cli/docker_cli_start_test.go [hunk 2]
```diff
@@ -47,12 +48,12 @@ func (s *DockerSuite) TestStartAttachCorrectExitCode(c *check.C) {
...
-	}
-	if exitCode != 1 {
-		c.Fatalf("start -a did not respond with proper exit code: expected 1, got %d", exitCode)
-	}
+	// start command should fail
+	c.Assert(err, checker.NotNil, check.Commentf("startOut: %s", startOut))
+	// start command failed unexpectedly
+	c.Assert(err.Error(), checker.Contains, "exit status 1", check.Commentf("startOut: %s", startOut))
+	// start -a did not respond with proper exit code
+	c.Assert(exitCode, checker.Equals, 1, check.Commentf("startOut: %s", startOut))
```
Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: Could you move the comment in the end of the line or in a `check.Commentf(…)` so we don't loose the "context" when this assertion fails ?

File: integration-cli/docker_cli_start_test.go [hunk 3]
```diff
@@ -47,12 +48,12 @@ func (s *DockerSuite) TestStartAttachCorrectExitCode(c *check.C) {
...
-	if err != nil && !strings.Contains("exit status 1", fmt.Sprintf("%s", err)) {
-		c.Fatalf("start command failed unexpectedly with error: %v, output: %q", err, startOut)
-	}
-	if exitCode != 1 {
-		c.Fatalf("start -a did not respond with proper exit code: expected 1, got %d", exitCode)
-	}
+	// start command should fail
+	c.Assert(err, checker.NotNil, check.Commentf("startOut: %s", startOut))
+	// start command failed unexpectedly
+	c.Assert(err.Error(), checker.Contains, "exit status 1", check.Commentf("startOut: %s", startOut))
```
Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: Could you move the comment in the end of the line or in a `check.Commentf(…)` so we don't loose the "context" when this assertion fails ?

File: integration-cli/docker_cli_start_test.go [hunk 4]
```diff
@@ -65,9 +66,8 @@ func (s *DockerSuite) TestStartAttachSilent(c *check.C) {
 	dockerCmd(c, "wait", name)
 
 	startOut, _ := dockerCmd(c, "start", "-a", name)
-	if expected := "test\n"; startOut != expected {
-		c.Fatalf("start -a produced unexpected output: expected %q, got %q", expected, startOut)
-	}
+	// start -a produced unexpected output
+	c.Assert(startOut, checker.Equals, "test\n")
```
Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: Could you move the comment in the end of the line or in a `check.Commentf(…)` so we don't loose the "context" when this assertion fails ?

File: integration-cli/docker_cli_start_test.go [hunk 5]
```diff
@@ -75,32 +75,26 @@ func (s *DockerSuite) TestStartRecordError(c *check.C) {
 	// when container runs successfully, we should not have state.Error
 	dockerCmd(c, "run", "-d", "-p", "9999:9999", "--name", "test", "busybox", "top")
 	stateErr, err := inspectField("test", "State.Error")
-	c.Assert(err, check.IsNil)
-	if stateErr != "" {
-		c.Fatalf("Expected to not have state error but got state.Error(%q)", stateErr)
-	}
+	c.Assert(err, checker.IsNil, check.Commentf("stateErr: %s", stateErr))
```
Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: Not sure why printing `stateErr` here ?_?

File: integration-cli/docker_cli_start_test.go [hunk 6]
```diff
@@ -75,32 +75,26 @@ func (s *DockerSuite) TestStartRecordError(c *check.C) {
...
+	c.Assert(err, checker.NotNil, check.Commentf("out: %s", out))
 
 	stateErr, err = inspectField("test2", "State.Error")
-	c.Assert(err, check.IsNil)
-	expected := "port is already allocated"
-	if stateErr == "" || !strings.Contains(stateErr, expected) {
-		c.Fatalf("State.Error(%q) does not include %q", stateErr, expected)
-	}
+	c.Assert(err, check.IsNil, check.Commentf("stateErr: %s", stateErr))
+	c.Assert(stateErr, checker.Contains, "port is already allocated")
```
Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: I would have kept the comment that was in `c.Fatalf` in a `check.Commentf`

File: integration-cli/docker_cli_start_test.go [hunk 7]
```diff
@@ -75,32 +75,26 @@ func (s *DockerSuite) TestStartRecordError(c *check.C) {
...
+	// Expected to not have state error
+	c.Assert(stateErr, checker.Equals, "")
 
 	// Expect this to fail and records error because of ports conflict
 	out, _, err := dockerCmdWithError("run", "-d", "--name", "test2", "-p", "9999:9999", "busybox", "top")
-	if err == nil {
-		c.Fatalf("Expected error but got none, output %q", out)
-	}
+	// Expected error but got none
+	c.Assert(err, checker.NotNil, check.Commentf("out: %s", out))
```
Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: I would have kept the comment that was in `c.Fatalf` in a `check.Commentf`

File: integration-cli/docker_cli_start_test.go [hunk 8]
```diff
@@ -111,9 +105,11 @@ func (s *DockerSuite) TestStartPausedContainer(c *check.C) {
...
 	dockerCmd(c, "pause", "testing")
 
-	if out, _, err := dockerCmdWithError("start", "testing"); err == nil || !strings.Contains(out, "Cannot start a paused container, try unpause instead.") {
-		c.Fatalf("an error should have been shown that you cannot start paused container: %s\n%v", out, err)
-	}
+	out, _, err := dockerCmdWithError("start", "testing")
+	// an error should have been shown that you cannot start paused container
+	c.Assert(err, checker.NotNil, check.Commentf("out: %s", out))
+	// an error should have been shown that you cannot start paused container
+	c.Assert(out, checker.Contains, "Cannot start a paused container, try unpause instead.")
```
Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: I would have kept the comment that was in `c.Fatalf` in a `check.Commentf`

File: integration-cli/docker_cli_start_test.go [hunk 9]
```diff
@@ -130,24 +126,23 @@ func (s *DockerSuite) TestStartMultipleContainers(c *check.C) {
...
 	// start all the three containers, container `child_first` start first which should be failed
 	// container 'parent' start second and then start container 'child_second'
 	out, _, err = dockerCmdWithError("start", "child_first", "parent", "child_second")
-	if !strings.Contains(out, "Cannot start container child_first") || err == nil {
-		c.Fatal("Expected error but got none")
-	}
+	// Expected error but got none
+	c.Assert(err, checker.NotNil)
+	// Expected error but got none
+	c.Assert(out, checker.Contains, "Cannot start container child_first")
```
Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: The comment could be better ; "output does not correspond to what was expected" or something like that :wink: .

File: integration-cli/docker_cli_start_test.go [hunk 10]
```diff
@@ -166,19 +161,18 @@ func (s *DockerSuite) TestStartAttachMultipleContainers(c *check.C) {
 	// test start and attach multiple containers at once, expected error
 	for _, option := range []string{"-a", "-i", "-ai"} {
 		out, _, err := dockerCmdWithError("start", option, "test1", "test2", "test3")
-		if !strings.Contains(out, "You cannot start and attach multiple containers at once.") || err == nil {
-			c.Fatal("Expected error but got none")
-		}
+		// Expected error but got none
+		c.Assert(err, checker.NotNil, check.Commentf("out: %s", out))
+		// Expected error but got none
+		c.Assert(out, checker.Contains, "You cannot start and attach multiple containers at once.")
```
Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: Same here

File: integration-cli/docker_cli_start_test.go (hunk 5 above)
Comment (WeiZhang555) [inline integration-cli/docker_cli_start_test.go]: Just don't want to lose any useful information, I just looked more deeply in the code, seems you are right, it's not necessary, will remove it.

File: integration-cli/docker_cli_start_test.go (hunk 1 above)
Comment (WeiZhang555) [inline integration-cli/docker_cli_start_test.go]: This is used to show why the assert failed, and is moved from `c.Fatal`.
I can find a better sentence.

File: integration-cli/docker_cli_start_test.go (hunk 7 above)
Comment (WeiZhang555) [inline integration-cli/docker_cli_start_test.go]: As we have discussed in #16854, comment above c.Assert statement will be printed out together with check.Commentf, do you still see a need to put comment in check.Commentf?
@vdemeester

Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: Nop, that's ok :wink:

**Comment by WeiZhang555** [comment]
@vdemeester updated per your suggestion. thank you! :smile:

File: integration-cli/docker_cli_start_test.go [hunk 11]
```diff
@@ -47,12 +48,12 @@ func (s *DockerSuite) TestStartAttachCorrectExitCode(c *check.C) {
...
 
 	startOut, exitCode, err := dockerCmdWithError("start", "-a", out)
-	if err != nil && !strings.Contains("exit status 1", fmt.Sprintf("%s", err)) {
-		c.Fatalf("start command failed unexpectedly with error: %v, output: %q", err, startOut)
-	}
-	if exitCode != 1 {
-		c.Fatalf("start -a did not respond with proper exit code: expected 1, got %d", exitCode)
-	}
+	// start command should fail
+	c.Assert(err, checker.NotNil, check.Commentf("startOut: %s", startOut))
```
Comment (HuKeping) [inline integration-cli/docker_cli_start_test.go]: `IsNil`?

Comment (WeiZhang555) [inline integration-cli/docker_cli_start_test.go]: Nop, should be `NotNil`, because `docker start` should fail with error.

Comment (vdemeester) [inline integration-cli/docker_cli_start_test.go]: @HuKeping I don't think so. The previous condition is not that clear but the test should fail, so `NotNil` is alright there :wink:.

Comment (HuKeping) [inline integration-cli/docker_cli_start_test.go]: Hmm, you  are right

**Comment by WeiZhang555** [comment]
Any progress on this?

**Comment by vdemeester** [comment]
LGTM 🐟

File: integration-cli/docker_cli_start_test.go [hunk 12]
```diff
@@ -15,9 +16,9 @@ func (s *DockerSuite) TestStartAttachReturnsOnError(c *check.C) {
 	dockerCmd(c, "wait", "test")
 
 	// Expect this to fail because the above container is stopped, this is what we want
-	if _, _, err := dockerCmdWithError("run", "-d", "--name", "test2", "--link", "test:test", "busybox"); err == nil {
-		c.Fatal("Expected error but got none")
-	}
+	_, _, err := dockerCmdWithError("run", "-d", "--name", "test2", "--link", "test:test", "busybox")
+	// err shouldn't be nil because container test2 try to link to stopped container
+	c.Assert(err, checker.NotNil)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_start_test.go]: Please include the output from `dockerCmdWithError` in a `check.Commentf`

File: integration-cli/docker_cli_start_test.go (hunk 3 above)
Comment (cpuguy83) [inline integration-cli/docker_cli_start_test.go]: This is a duplicate of checking the exit code below. We can remove it.

File: integration-cli/docker_cli_start_test.go [hunk 13]
```diff
@@ -130,24 +126,23 @@ func (s *DockerSuite) TestStartMultipleContainers(c *check.C) {
...
+	c.Assert(out, checker.Equals, "false")
 
 	// start all the three containers, container `child_first` start first which should be failed
 	// container 'parent' start second and then start container 'child_second'
 	out, _, err = dockerCmdWithError("start", "child_first", "parent", "child_second")
-	if !strings.Contains(out, "Cannot start container child_first") || err == nil {
-		c.Fatal("Expected error but got none")
-	}
+	// err shouldn't be nil because start will fail
+	c.Assert(err, checker.NotNil)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_start_test.go]: Please include the `out` in a check.Commentf

File: integration-cli/docker_cli_start_test.go [hunk 14]
```diff
@@ -130,24 +126,23 @@ func (s *DockerSuite) TestStartMultipleContainers(c *check.C) {
...
+	// output does not correspond to what was expected
+	c.Assert(out, checker.Contains, "Cannot start container child_first")
 
 	for container, expected := range map[string]string{"parent": "true", "child_first": "false", "child_second": "true"} {
 		out, err := inspectField(container, "State.Running")
+		// Container running state wrong
 		c.Assert(err, check.IsNil)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_start_test.go]: Can you have this include `out` in `check.Commentf` as well?

File: integration-cli/docker_cli_start_test.go [hunk 15]
```diff
@@ -166,19 +161,18 @@ func (s *DockerSuite) TestStartAttachMultipleContainers(c *check.C) {
...
 	for container, expected := range map[string]string{"test1": "false", "test2": "false", "test3": "false"} {
 		out, err := inspectField(container, "State.Running")
-		if err != nil {
-			c.Fatal(out, err)
-		}
-		if out != expected {
-			c.Fatal("Container running state wrong")
-		}
+		// Container running state wrong
+		c.Assert(err, check.IsNil)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_start_test.go]: include `out` please :)

**Comment by WeiZhang555** [comment]
@cpuguy83 updated per your suggestion, thank you! :)

File: integration-cli/docker_cli_start_test.go [hunk 16]
```diff
@@ -65,9 +64,8 @@ func (s *DockerSuite) TestStartAttachSilent(c *check.C) {
 	dockerCmd(c, "wait", name)
 
 	startOut, _ := dockerCmd(c, "start", "-a", name)
-	if expected := "test\n"; startOut != expected {
-		c.Fatalf("start -a produced unexpected output: expected %q, got %q", expected, startOut)
-	}
+	// start -a produced unexpected output
+	c.Assert(startOut, checker.Equals, "test\n")
```
Comment (cpuguy83) [inline integration-cli/docker_cli_start_test.go]: This one still needs a check.Commentf for `startOut`

File: integration-cli/docker_cli_start_test.go [hunk 17]
```diff
@@ -75,32 +73,26 @@ func (s *DockerSuite) TestStartRecordError(c *check.C) {
 	// when container runs successfully, we should not have state.Error
 	dockerCmd(c, "run", "-d", "-p", "9999:9999", "--name", "test", "busybox", "top")
 	stateErr, err := inspectField("test", "State.Error")
-	c.Assert(err, check.IsNil)
-	if stateErr != "" {
-		c.Fatalf("Expected to not have state error but got state.Error(%q)", stateErr)
-	}
+	c.Assert(err, checker.IsNil)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_start_test.go]: This one still needs a check.Commentf for `stateErr`

File: integration-cli/docker_cli_start_test.go [hunk 18]
```diff
@@ -75,32 +73,26 @@ func (s *DockerSuite) TestStartRecordError(c *check.C) {
...
 	// Expect this to fail and records error because of ports conflict
 	out, _, err := dockerCmdWithError("run", "-d", "--name", "test2", "-p", "9999:9999", "busybox", "top")
-	if err == nil {
-		c.Fatalf("Expected error but got none, output %q", out)
-	}
+	// err shouldn't be nil because docker run will fail
+	c.Assert(err, checker.NotNil, check.Commentf("out: %s", out))
 
 	stateErr, err = inspectField("test2", "State.Error")
 	c.Assert(err, check.IsNil)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_start_test.go]: This one still needs a check.Commentf for `startOut`

File: integration-cli/docker_cli_start_test.go [hunk 19]
```diff
@@ -75,32 +73,26 @@ func (s *DockerSuite) TestStartRecordError(c *check.C) {
...
-	if stateErr == "" || !strings.Contains(stateErr, expected) {
-		c.Fatalf("State.Error(%q) does not include %q", stateErr, expected)
-	}
+	c.Assert(stateErr, checker.Contains, "port is already allocated")
 
 	// Expect the conflict to be resolved when we stop the initial container
 	dockerCmd(c, "stop", "test")
 	dockerCmd(c, "start", "test2")
 	stateErr, err = inspectField("test2", "State.Error")
 	c.Assert(err, check.IsNil)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_start_test.go]: This one still needs a check.Commentf for `startErr`

File: integration-cli/docker_cli_start_test.go [hunk 20]
```diff
@@ -130,24 +124,23 @@ func (s *DockerSuite) TestStartMultipleContainers(c *check.C) {
 
 	out, err := inspectField("parent", "State.Running")
 	c.Assert(err, check.IsNil)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_start_test.go]: Include out

**Comment by cpuguy83** [comment]
Sorry, added a few more places where we need output, I know some of them you didn't touch, but would be super helpful if you could add those in.

Thanks!

File: integration-cli/docker_cli_start_test.go (hunk 16 above)
Comment (WeiZhang555) [inline integration-cli/docker_cli_start_test.go]: Why? Won't `Assert` print `startOut` for us?

**Comment by WeiZhang555** [comment]
All right, I will do this.

File: integration-cli/docker_cli_start_test.go (hunk 16 above)
Comment (cpuguy83) [inline integration-cli/docker_cli_start_test.go]: Yep, you're right.

**Comment by WeiZhang555** [comment]
@cpuguy83 updated. :)

**Comment by cpuguy83** [comment]
LGTM

**Pull request merged**